File: facebook/src/com/facebook/WorkQueue.java
package com.facebook;

import java.util.concurrent.Executor;

class WorkQueue {
    public static final int DEFAULT_MAX_CONCURRENT = 8;

    private Object pendingJobsLock = new Object();
    private Object runningJobsLock = new Object();
    private WorkNode pendingJobs;

    private final int maxConcurrent;
    private final Executor executor;

    private WorkNode runningJobs = null;
    private int runningCount = 0;

    WorkQueue() {
        this(DEFAULT_MAX_CONCURRENT);
    }

    WorkQueue(int maxConcurrent) {
        this(maxConcurrent, Settings.getExecutor());
    }

    WorkQueue(int maxConcurrent, Executor executor) {
        this.maxConcurrent = maxConcurrent;
        this.executor = executor;
    }

    WorkItem addActiveWorkItem(Runnable callback) {
        return addActiveWorkItem(callback, true);
    }

    WorkItem addActiveWorkItem(Runnable callback, boolean addToFront) {
        WorkNode node = new WorkNode(callback);
        synchronized (pendingJobsLock) {
            pendingJobs = node.addToList(pendingJobs, addToFront);
        }

        startItem();
        return node;
    }

    void validate() {
        synchronized (runningJobsLock) {
            // Verify that all running items know they are running, and counts match
            int count = 0;

            if (runningJobs != null) {
                WorkNode walk = runningJobs;
                do {
                    walk.verify(true);
                    count++;
                    walk = walk.getNext();
                } while (walk != runningJobs);
            }

            assert runningCount == count;
        }
    }

    private void startItem() {
        finishItemAndStartNew(null);
    }

    private void finishItemAndStartNew(WorkNode finished) {
        WorkNode ready = null;

        synchronized (runningJobsLock) {
            if (finished != null) {
                runningJobs = finished.removeFromList(runningJobs);
                runningCount--;
            }

            if (runningCount < maxConcurrent) {
                ready = extractNextReadyItem();

                if (ready != null) {
                    runningJobs = ready.addToList(runningJobs, false);
                    runningCount++;
                }
            }
        }

        if (ready != null) {
            execute(ready);
        }
    }

    private void execute(final WorkNode node) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    node.getCallback().run();
                } finally {
                    finishItemAndStartNew(node);
                }
            }
        });
    }

    private WorkNode extractNextReadyItem() {
        synchronized (pendingJobsLock) {
            WorkNode ready = pendingJobs;
            if (ready != null) {
                pendingJobs = ready.removeFromList(pendingJobs);
                ready.setIsRunning(true);
                return ready;
            }
        }

        return null;
    }

    private class WorkNode implements WorkItem {
        private final Runnable callback;
        private WorkNode next;
        private WorkNode prev;
        private boolean isRunning;

        WorkNode(Runnable callback) {
            this.callback = callback;
        }

        @Override
        public boolean cancel() {
            synchronized (pendingJobsLock) {
                if (!isRunning()) {
                    pendingJobs = removeFromList(pendingJobs);
                    return true;
                }
            }

            return false;
        }

        @Override
        public void moveToFront() {
            synchronized (pendingJobsLock) {
                if (!isRunning()) {
                    pendingJobs = removeFromList(pendingJobs);
                    pendingJobs = addToList(pendingJobs, true);
                }
            }
        }

        @Override
        public boolean isRunning() {
            return isRunning;
        }

        Runnable getCallback() {
            return callback;
        }

        WorkNode getNext() {
            return next;
        }

        void setIsRunning(boolean isRunning) {
            this.isRunning = isRunning;
        }

        WorkNode addToList(WorkNode list, boolean addToFront) {
            assert next == null;
            assert prev == null;

            if (list == null) {
                list = next = prev = this;
            } else {
                next = list;
                prev = list.prev;
                next.prev = prev.next = this;
            }

            return addToFront ? this : list;
        }

        WorkNode removeFromList(WorkNode list) {
            assert next != null;
            assert prev != null;

            if (list == this) {
                if (next == this) {
                    list = null;
                } else {
                    list = next;
                }
            }

            next.prev = prev;
            prev.next = next;
            next = prev = null;

            return list;
        }

        void verify(boolean shouldBeRunning) {
            assert prev.next == this;
            assert next.prev == this;
            assert isRunning() == shouldBeRunning;
        }
    }

    interface WorkItem {
        boolean cancel();
        boolean isRunning();
        void moveToFront();
    }
}

File: facebook-core/src/main/java/com/facebook/internal/WorkQueue.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal

import com.facebook.FacebookException
import com.facebook.FacebookSdk
import java.util.concurrent.Executor
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
class WorkQueue
@JvmOverloads
constructor(
    private val maxConcurrent: Int = DEFAULT_MAX_CONCURRENT,
    private val executor: Executor = FacebookSdk.getExecutor()
) {
  private val workLock = ReentrantLock()
  private var pendingJobs: WorkNode? = null
  private var runningJobs: WorkNode? = null
  private var runningCount = 0

  companion object {
    const val DEFAULT_MAX_CONCURRENT = 8
    private fun assert(condition: Boolean) {
      if (!condition) {
        throw FacebookException("Validation failed")
      }
    }
  }

  @JvmOverloads
  fun addActiveWorkItem(callback: Runnable, addToFront: Boolean = true): WorkItem {
    val node = WorkNode(callback)
    workLock.withLock { pendingJobs = node.addToList(pendingJobs, addToFront) }
    startItem()
    return node
  }

  fun validate() {
    // Verify that all running items know they are running, and counts match
    workLock.withLock {
      var count = 0
      if (runningJobs != null) {
        var walk = runningJobs
        do {
          checkNotNull(walk)
          walk.verify(true)
          count++
          walk = walk.next
        } while (walk !== runningJobs)
      }
      assert(runningCount == count)
    }
  }

  private fun startItem() {
    finishItemAndStartNew(null)
  }

  private fun finishItemAndStartNew(finished: WorkNode?) {
    var ready: WorkNode? = null
    workLock.lock()

    if (finished != null) {
      runningJobs = finished.removeFromList(runningJobs)
      runningCount--
    }
    if (runningCount < maxConcurrent) {
      ready = pendingJobs // Head of the pendingJobs queue
      if (ready != null) {
        // The Queue reassignments are necessary since 'ready' might have been
        // added / removed from the front of either queue, which changes its
        // respective head.
        pendingJobs = ready.removeFromList(pendingJobs)
        runningJobs = ready.addToList(runningJobs, false)
        runningCount++
        ready.isRunning = true
      }
    }
    workLock.unlock()
    if (ready != null) {
      execute(ready)
    }
  }

  private fun execute(node: WorkNode) {
    executor.execute {
      try {
        node.callback.run()
      } finally {
        finishItemAndStartNew(node)
      }
    }
  }

  private inner class WorkNode constructor(val callback: Runnable) : WorkItem {
    var next: WorkNode? = null
      private set
    private var prev: WorkNode? = null
    override var isRunning = false
    override fun cancel(): Boolean {
      workLock.withLock {
        if (!isRunning) {
          pendingJobs = removeFromList(pendingJobs)
          return true
        }
      }
      return false
    }

    override fun moveToFront() {
      workLock.withLock {
        if (!isRunning) {
          pendingJobs = removeFromList(pendingJobs)
          pendingJobs = addToList(pendingJobs, true)
        }
      }
    }

    fun addToList(list: WorkNode?, addToFront: Boolean): WorkNode {
      var list = list
      assert(next == null)
      assert(prev == null)
      if (list == null) {
        prev = this
        next = prev
        list = next
      } else {
        next = list
        prev = list.prev
        prev?.next = this
        next?.prev = prev?.next
      }
      checkNotNull(list)
      return if (addToFront) this else list
    }

    fun removeFromList(list: WorkNode?): WorkNode? {
      var list = list
      assert(next != null)
      assert(prev != null)
      if (list === this) {
        list =
            if (next === this) {
              null
            } else {
              next
            }
      }
      next?.prev = prev
      prev?.next = next
      prev = null
      next = prev
      return list
    }

    fun verify(shouldBeRunning: Boolean) {
      assert(prev?.next ?: this === this)
      assert(next?.prev ?: this === this)
      assert(isRunning == shouldBeRunning)
    }
  }

  interface WorkItem {
    fun cancel(): Boolean
    val isRunning: Boolean
    fun moveToFront()
  }
}
